integrations: Add Constructor Groups video call provider

fatih-constructor-groups opened the pull request:
This PR adds Constructor Groups video calling integration to Zulip, allowing users to create video calls using the Constructor Groups service.

## Changes Made
- **Backend Integration**: Added `ConstructorGroupsService` class that uses a single API endpoint (`POST /room/default`) to get or create video call rooms
- **Frontend Components**: Added video call creation buttons and proper XHR handling with abort logic
- **Configuration**: Added settings for Constructor Groups URL, access key, and secret key
- **Error Handling**: Added comprehensive error handling with proper user feedback
- **Testing**: Added complete test coverage for both backend API and frontend UI interactions
- **Documentation**: Added help documentation, integration templates, changelog entry, and updated OpenAPI specs

## Key Features
- **Simplified API Integration**: Uses single endpoint that handles both room lookup and creation atomically
- **Smart Room Naming**: Uses user's full name with realm_id fallback for uniqueness across organizations
- **Defensive Programming**: Handles concurrent clicks with proper XHR abort functionality
- **Audio Call Hiding**: Hides audio call button for Constructor Groups provider as requested

Fixes: #35623

**Screenshots and screen captures:**

### 1. Organization Settings
<img width="946" height="850" alt="call-provider" src="https://github.com/user-attachments/assets/a4c3d42f-2884-4281-b7d1-3056a00ba6fa" />

### 2. Integration Documentation  
<img width="1578" height="986" alt="image" src="https://github.com/user-attachments/assets/8b191e12-1f23-4a22-8ebf-4f788b92256f" />

### 3. Video Call Feature in Action
<img width="1585" height="989" alt="image" src="https://github.com/user-attachments/assets/c158a151-d793-41df-aef4-301b7753ae91" />

### 4. Error Message Handling
<img width="942" height="858" alt="failed-video-call" src="https://github.com/user-attachments/assets/31c36ee8-4612-4e4d-838b-4e847bf33424" />

### 5. API Documentation
<img width="1000" height="995" alt="image" src="https://github.com/user-attachments/assets/c1c99e4f-d4f1-4671-92a5-f721971c3113" />

### 6. Video call providers
<img width="1005" height="990" alt="image" src="https://github.com/user-attachments/assets/9401e07d-a34c-4db9-b6e8-c845565c533d" />

### 7. Start a call
<img width="1284" height="996" alt="image" src="https://github.com/user-attachments/assets/7632900d-f7af-476c-a0f1-835857c6bcbb" />

### 8. Configure call provider
<img width="1290" height="951" alt="image" src="https://github.com/user-attachments/assets/21363e7e-c5ea-4218-9abd-cc69a9301bc0" />


<details>
<summary>Self-review checklist</summary>

- [x] Self-reviewed the changes for clarity and maintainability
- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.
- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.
- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

fatih-constructor-groups commented on help/include/sidebar_index.md:
```diff
@@ -85,6 +85,7 @@
 * [Collaborative to-do lists](/help/collaborative-to-do-lists)
 * [Global times](/help/global-times)
 * [Start a call](/help/start-a-call)
+* [Constructor Groups calls](/help/constructor-groups-calls)
```
Removed `help/constructor-groups-calls.md` file entirely

fatih-constructor-groups commented on zerver/views/video_calls.py:
```diff
@@ -47,6 +48,75 @@ def __init__(self) -> None:
+
+        return self._make_authenticated_request("POST", "/room", data)
+
+    def get_room_url(self, room_data: dict[str, Any]) -> str:
```
Done! Inlined the function.

fatih-constructor-groups commented on zerver/views/video_calls.py:
```diff
@@ -406,3 +483,45 @@ def join_bigbluebutton(request: HttpRequest, *, bigbluebutton: str) -> HttpRespo
+                raise JsonableError(_("Invalid room data received from Constructor Groups."))
+        else:
+            # Create new room with standard naming convention
+            room_name = f"{user_profile.full_name}'s Zulip Videoconferencing Room"
```
Reused `room_name_pattern` instead of creating redundant `room_name` variable

fatih-constructor-groups commented on zerver/views/video_calls.py:
```diff
@@ -406,3 +483,45 @@ def join_bigbluebutton(request: HttpRequest, *, bigbluebutton: str) -> HttpRespo
+        existing_rooms = service.search_rooms(room_name_pattern, user_profile.delivery_email)
+
+        if existing_rooms:
+            # Use existing room
```
Removed all self-explanatory comments from the function

fatih-constructor-groups commented on zerver/views/video_calls.py:
```diff
@@ -406,3 +483,45 @@ def join_bigbluebutton(request: HttpRequest, *, bigbluebutton: str) -> HttpRespo
+        raise
+    except Exception:
+        # Log the full error for debugging but show user-friendly message
+        logging.exception("Constructor Groups API error")
```
Removed the error logging to match other video call integrations

fatih-constructor-groups commented on zerver/views/video_calls.py:
```diff
@@ -406,3 +483,45 @@ def join_bigbluebutton(request: HttpRequest, *, bigbluebutton: str) -> HttpRespo
+    except Exception:
+        # Log the full error for debugging but show user-friendly message
+        logging.exception("Constructor Groups API error")
+        raise JsonableError(_("Failed to create video call"))
```
Removed the entire try-except block since errors are handled internally in the class

fatih-constructor-groups commented on docs/production/video-calls.md:
```diff
@@ -161,3 +162,32 @@ This enables BigBlueButton support in your Zulip server. Finally, [configure
 provider](https://zulip.com/help/configure-call-provider)
 in the Zulip organizations where you want to use it.
+
+## Constructor Groups
```
Removed both duplicate sections.

fatih-constructor-groups commented:
@Niloth-p Thanks for the detailed feedback and thorough review! 

## Tests
**Updated existing test files instead of creating new ones:**
- Updated `test_create_video_call.py` with Constructor Groups test cases
- Updated `compose_video.test.cjs` with frontend tests
- Updated `test_realm.py` with realm setting tests
- Updated `settings_org.test.cjs` with organization settings tests
- Updated `test_extra_settings.py` with configuration tests
- Removed standalone `test_constructor_groups.py` file

## Other files to update
**Completed:**
- **zulip.yaml**: Added Constructor Groups endpoint documentation
- **rest-endpoints.md**: Added Constructor Groups to API endpoints list

 **Not needed:**
- **comparison_table_integrated.html**: Contains plan comparisons, not video providers
- **security-model.md**: Covers general architecture, not provider-specific security

Constructor Groups follows same patterns as other video providers.

## Audio call support  
**Decision: No audio call support** - implemented your second option:

**Backend changes:**
- Removed `is_video_call` parameter from `make_constructor_groups_video_call`
- Changed to `@typed_endpoint_without_parameters`

**Frontend changes:**  
- Removed `is_video_call`/`is_audio_call` handling in `compose_call_ui.ts`
- Now sends empty data `{}` for video calls only

**Audio button:** Already properly hidden (not in `compute_show_audio_chat_button`)

## Commits
All linting issues have been addressed in the latest commit.

## Screenshots  
Will add screenshots to PR description showing:
- New video call feature in action
- Error message handling  

Ready for another review! 🚀

fatih-constructor-groups commented on templates/zerver/integrations/constructor-groups.md:
```diff
@@ -0,0 +1,49 @@
+
+## Configure Constructor Groups as your call provider
+
+By default, Zulip integrates with
```
added

fatih-constructor-groups commented on templates/zerver/integrations/constructor-groups.md:
```diff
@@ -0,0 +1,49 @@
+!!! warn ""
+
+    **Note:** This is currently only possible on self-hosted Zulip
+    installations, and you'll need Constructor Groups API access.
```
I've updated it to "you'll need a Constructor Groups account" to be more accurate and consistent with other integrations.
The documentation link already covers how to create API credentials once you have an account.

fatih-constructor-groups commented on zproject/prod_settings_template.py:
```diff
@@ -733,6 +733,10 @@
 
+## Controls the Constructor Groups video call integration. You must also
+## set constructor_groups_access_key and constructor_groups_secret_key in zulip-secrets.conf.
+# CONSTRUCTOR_GROUPS_URL = "https://constructor.app/api/groups/xapi"
```
I've updated it to be a template URL: `https://yourgroupsdomain.com/api/groups/xapi`

fatih-constructor-groups commented on zerver/tests/test_create_video_call.py:
```diff
@@ -533,3 +533,167 @@ def test_join_bigbluebutton_redirect_not_configured(self) -> None:
+    @mock.patch("zerver.views.video_calls.settings.CONSTRUCTOR_GROUPS_SECRET_KEY", None)
+    def test_not_configured_secret_key(self) -> None:
+        response = self.client_post("/json/calls/constructorgroups/create")
+        self.assert_json_error(response, "Failed to create video call")
```
Done

fatih-constructor-groups commented on zerver/tests/test_create_video_call.py:
```diff
@@ -533,3 +533,167 @@ def test_join_bigbluebutton_redirect_not_configured(self) -> None:
+        CONSTRUCTOR_GROUPS_SECRET_KEY="test-secret-key",
+    )
+    @mock.patch("zerver.views.video_calls.ConstructorGroupsService._make_authenticated_request")
+    def test_call_creation_no_parameter_distinction(self, mock_request: mock.Mock) -> None:
```
I've removed

fatih-constructor-groups commented on zerver/tests/test_create_video_call.py:
```diff
@@ -533,3 +533,167 @@ def test_join_bigbluebutton_redirect_not_configured(self) -> None:
+        super().setUp()
+        self.user_profile = self.example_user("hamlet")
+        self.login_user(self.user_profile)
+
```
Done

fatih-constructor-groups commented on web/src/compose_call.ts:
```diff
@@ -11,7 +11,10 @@ export function abort_video_callbacks(edit_message_id = ""): void {
     const xhr = video_call_xhrs.get(edit_message_id);
     if (xhr !== undefined) {
-        xhr.abort();
+        // Safely call abort if it exists
```
@Niloth-p This defensive check is essential to prevent crashes from rapid button clicks. Without it, when users quickly click video call buttons multiple times, `xhr.abort()` can throw `TypeError: xhr.abort is not a function` if the XHR object doesn't have a proper abort method or if there are race conditions.

This is a user-facing production issue, not just a testing concern. The Constructor Groups integration uses the same XHR tracking pattern as Zoom, so both need this protection. I've verified this prevents crashes during rapid clicking scenarios.

fatih-constructor-groups replied:
@Niloth-p That code block has been reverted to its original state. The defensive check is now only in our Constructor Groups section (line 176).

fatih-constructor-groups commented:
@Niloth-p Thank you so much for all your detailed feedback throughout this PR! 🙏 

I really appreciate the time you've taken to go through everything.

**Final request completed:** Removed the docstrings and self-explanatory comments from the five latest `_make_authenticated_request` tests as requested.

Thanks again for the excellent collaboration! 🚀

alya commented:
@laurynmm @Niloth-p Is this PR ready for integration review from your perspective?

laurynmm commented:
@fatih-constructor-groups - Thanks for the quick response to those questions!

Sounds like question one shouldn't be an issue as the `creator_email` will be set to the Zulip user's email.

For question two, in Zulip's API, the translated [message strings in errors](https://zulip.com/api/rest-error-handling) aren't intended to be used for debugging. They are intended to be to used by clients for presenting human-readable error information.

Since it sounds like the client (e.g., the web app) and the Zulip user can't do anything for those errors, we might want to instead have a logging on the server side for debugging and send the client the same error message for all Constructor Group server side errors. Maybe instead of "Failed to create video call", something more specific like "Failed to create Constructor Group video call room."? I'll let @timabbott weigh in on what might be best.

fatih-constructor-groups commented:
@laurynmm I've made the requested changes:

## Changes Made

- **Simplified error handling**: Replaced the two specific error messages with a single, message: "Failed to create Constructor Group video call room."
- **Updated tests**: Modified the test cases to expect the new unified error message.

## Additional Improvements

I also added **room name conflict handling**:
- If room name conflicts occur (409 error), the system automatically retries with a user ID suffix
- Example: "John's Zulip Videoconferencing Room" → "John-123's Zulip Videoconferencing Room"
- This prevents room creation failures due to duplicate names

## Documentation Updates

Updated the integration documentation to explain room naming.

Ready for the next review! 🚀

fatih-constructor-groups commented on zerver/views/video_calls.py:
```diff
@@ -406,3 +485,38 @@ def join_bigbluebutton(request: HttpRequest, *, bigbluebutton: str) -> HttpRespo
+                fallback_name = (
+                    f"{user_profile.full_name}-{user_profile.id}'s Zulip Videoconferencing Room"
+                )
+                room_data = service.create_room(fallback_name, user_profile.delivery_email)
```
You're right that the second call doesn't have explicit error handling, but this is intentional. The fallback name is only used when there's a room name conflict (409 error), and if the fallback name creation also fails, the user will receive a notification through the existing error handling mechanism.

fatih-constructor-groups commented on web/src/compose_call_ui.ts:
```diff
@@ -145,6 +145,39 @@ export function generate_and_insert_audio_or_video_call_link(
+            error(xhr, status) {
+                compose_call.video_call_xhrs.delete(key);
+                if (status !== "abort") {
+                    const parsed = z.object({msg: z.string()}).safeParse(xhr.responseJSON);
+                    if (parsed.success && parsed.data.msg.includes("Room name conflict")) {
+                        return;
+                    }
```
This handles the case where the backend returns a 409 error with 'Room name conflict' message but doesn't re-raise it as an exception.

fatih-constructor-groups commented:
> Vay canına, ne kadar çok değişiklik olmuş. Zaman ayırıp emek verdiğiniz için teşekkürler! İsim çakışmaları için destek eklemek gerçekten güzel. Yeni parametre `is_zulip_room`faydalı görünüyor. Testlerin çoğunu yeniden yazdığınızı `test_create_video_call.py`ve `_make_authenticated_request`.
> 
> Aşağıda birkaç düşüncemi paylaştım. Bazı gereksiz yorumlar geri eklendi. Bana göre iyi görünüyor, güncellemeler için teşekkür ederim.

@Niloth-p  Thanks for the review! 😊

**Updates made:**
- `is_zulip_room` → `is_default_room` 
- Removed redundant comment
- Fixed method descriptions

Appreciate your time reviewing all these changes!

Niloth-p commented:
LGTM, thanks for updating your API!
Sorry for the delay in getting back.
Thanks for updating the example URL as well.
Let's just remove the comments that have been re-introduced. I've attached the diff you can patch.

```diff
diff --git a/templates/zerver/integrations/constructor-groups.md b/templates/zerver/i
ntegrations/constructor-groups.md
index 55d1992de4..3407d5b972 100644
--- a/templates/zerver/integrations/constructor-groups.md
+++ b/templates/zerver/integrations/constructor-groups.md
@@ -48,10 +48,6 @@ source video conferencing solution. You can configure Zulip t
o use
 
 {end_tabs}
 
-## Room naming and conflicts
-
-Constructor Groups video call rooms are created with names based on the user's full
 name. If multiple users in your organization have the same full name, room name conflic
ts may occur. When a conflict occurs, the room name will include the user ID to ensure u
niqueness (e.g., **John-123's Zulip Videoconferencing Room**).
-
 ### Related documentation
 
 - [Constructor Groups API documentation][constructor-groups-api-docs]
diff --git a/zerver/tests/test_create_video_call.py b/zerver/tests/test_create_video_
call.py
index 7749c22cdf..50bd94a8ac 100644
--- a/zerver/tests/test_create_video_call.py
+++ b/zerver/tests/test_create_video_call.py
@@ -573,7 +573,6 @@ class ConstructorGroupsVideoCallTest(ZulipTestCase):
 
     @responses.activate
     def test_get_existing_room(self) -> None:
-        """Test that existing room is returned with 200 status"""
         responses.add(
             responses.POST,
             f"{self.base_api_url}/room/default",
@@ -589,7 +588,6 @@ class ConstructorGroupsVideoCallTest(ZulipTestCase):
 
     @responses.activate
     def test_create_new_room(self) -> None:
-        """Test that new room is created with 201 status"""
         responses.add(
             responses.POST,
             f"{self.base_api_url}/room/default",
@@ -605,7 +603,6 @@ class ConstructorGroupsVideoCallTest(ZulipTestCase):
 
     @responses.activate
     def test_missing_url_in_response(self) -> None:
-        """Test handling when API returns room data without URL"""
         responses.add(
             responses.POST,
             f"{self.base_api_url}/room/default",
@@ -621,7 +618,6 @@ class ConstructorGroupsVideoCallTest(ZulipTestCase):
 
     @responses.activate
     def test_http_request_error_handling(self) -> None:
-        """Test handling when API returns 500 error"""
         responses.add(
             responses.POST,
             f"{self.base_api_url}/room/default",
@@ -635,7 +631,6 @@ class ConstructorGroupsVideoCallTest(ZulipTestCase):
 
     @responses.activate
     def test_api_response_json_error(self) -> None:
-        """Test handling when API returns invalid JSON"""
         responses.add(
             responses.POST,
             f"{self.base_api_url}/room/default",
@@ -650,7 +645,6 @@ class ConstructorGroupsVideoCallTest(ZulipTestCase):
 
     @responses.activate
     def test_request_contains_correct_data(self) -> None:
-        """Test that request contains correct room name and fallback name"""
         responses.add(
             responses.POST,
             f"{self.base_api_url}/room/default",
@@ -674,7 +668,6 @@ class ConstructorGroupsVideoCallTest(ZulipTestCase):
         )
 
     def test_unsupported_http_method(self) -> None:
-        """Test that unsupported HTTP methods raise an error"""
         from zerver.lib.exceptions import JsonableError
         from zerver.views.video_calls import ConstructorGroupsService
 
diff --git a/zerver/views/video_calls.py b/zerver/views/video_calls.py
index b8d31bbc5a..97a31e32a9 100644
--- a/zerver/views/video_calls.py
+++ b/zerver/views/video_calls.py
@@ -106,11 +106,6 @@ class ConstructorGroupsService:
     def get_or_create_default_room(
         self, creator_email: str, name: str, fallback_name: str
     ) -> dict[str, Any]:
-        """Get existing default room or create new one if not exists.
-
-        Uses single API call to /room/default endpoint.
-        Returns 200 if room already exists, 201 if newly created.
-        """
         data = {
             "creator_email": creator_email,
             "name": name,
```

zafergurel commented:
Thank you, everyone, for the effort.
